internal/util/recoverable.go

```diff
@@ -0,0 +1,113 @@
+package util
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"runtime/debug"
+	"sync"
+	"time"
+)
+
+var (
+	ErrServiceStopped = fmt.Errorf("service stopped")
+)
+
+type Doable interface {
+	Do() error
+	Stop()
+}
+
+func NewRecoverableService(svc Doable, logger *log.Logger) *RecoverableService {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &RecoverableService{
+		service: svc,
+		stopped: make(chan stoppedService, 1),
+		log:     logger,
+		ctx:     ctx,
+		cancel:  cancel,
+	}
+}
+
+type RecoverableService struct {
+	mu      sync.Mutex
+	running bool
+	service Doable
+	stopped chan stoppedService
```

Comment:
Why is stopped a `chan stoppedService` and not just a `chan error`? I might be missing something but it looks like when we read a `stoppedService` off the channel, we only want to access the error it wraps?

Reply:
To be honest, I don't know why I did that. You're right though. No need for the wrapper.